superduperdb/components/graph.py

```diff
@@ -0,0 +1,267 @@
+        G_.add_edge(u, v, weight=on or self._DEFAULT_ARG_WEIGHT)
+
+        if not nx.is_directed_acyclic_graph(G_):
+            raise TypeError('The graph is not DAG with this edge')
+        self.G = G_
+        return self.nodes[v]
+
+    def stash_node_output(self, node, output):
+        self._node_output_cache[node.name] = output
+
+    def group_nodes_by_degree(self, nodes):
+        grouped_dict = OrderedDict()
+        for item in nodes:
+            key = item[1]
+            value = item[0]
+            if key not in grouped_dict:
+                grouped_dict[key] = []
+            grouped_dict[key].append(value)
+        return grouped_dict
+
+    def level_traversal(self, G, nodes, traversal_path=[]):
+        if len(nodes) == 1:
+            traversal_path.append(nodes[0])
+            return traversal_path
+        G = G.subgraph(nodes)
+
+        # Traverse
+        nodes = sorted(G.in_degree, key=lambda x: x[1], reverse=False)
+        grouped_nodes = self.group_nodes_by_degree(nodes)
+
+        # Find zero in bound degree nodes and add to tranversal_path
+        zero_bound_nodes = grouped_nodes.pop(0)
+        _ = [traversal_path.append(n) for n in zero_bound_nodes]
+
+        for _, nodes in grouped_nodes.items():
+            self.level_traversal(G, nodes, traversal_path)
+        return traversal_path
+
+    def traversal(self, graph, nodes, traversal_path=[]):
+        if not nodes:
+            return traversal_path
+
+        def find_level_neighbors(graph, nodes):
+            neighbors = []
+
+            for node in nodes:
+                neighbor = list(graph.neighbors(node))
+                if neighbor:
+                    _ = [neighbors.append(n) for n in neighbor]
+            if neighbors:
+                neighbors = set(neighbors) - set(nodes)
+                neighbors = list(neighbors)
+            return neighbors
+
+        neighbors = find_level_neighbors(graph, nodes)
+
+        if not neighbors:
+            traversal_path += nodes
+            return traversal_path
+
+        S = graph.subgraph(neighbors)
+        traversal_path = self.level_traversal(S, neighbors, traversal_path)
+
+        neighbors = find_level_neighbors(graph, neighbors)
+        return self.traversal(graph, neighbors, traversal_path)
+
+    def validate(self, path):
+        for node in path:
+            nodes = list(self.G.predecessors(node))
+
+            arg_nodes = list(map(lambda x: self.nodes[x], nodes))
+            node = self.nodes[node]
+
+            if node.required_input != len(arg_nodes):
+                raise TypeError(
+                    f'Graph disconnected at Node: {node.name} '
+                    f'and is partially connected with {nodes}\n'
+                    f'Required connected node is {node.required_input} '
+                    f'but got only {len(nodes)}, '
+                    f'Node required params: {node.params}'
+                )
+
+    def predict(self, x, one=True, select=None):
```

Comment:
Should be `self._predict`, then the select logic isn't necessary.

Reply:
done